purpleair-aqi.js

```diff
@@ -41,15 +42,93 @@ const SENSOR_ID = args.widgetParameter || "69223";
  * @property {string} lon
  */
 
+/**
+ * @typedef {object} LatLon
+ * @property {number} latitude
+ * @property {number} longitude
+ */
+
+/**
+ * Get the closest PurpleAir sensorId to the given location
+ *
+ * @param {LatLon} location
+ * @returns {Promise<number>}
+ */
+async function getSensorId({ latitude, longitude }) {
+  if (SENSOR_ID) return SENSOR_ID;
+
+  const BOUND_OFFSET = 0.05;
+
+  const nwLat = latitude + BOUND_OFFSET;
+  const seLat = latitude - BOUND_OFFSET;
+  const nwLng = longitude - BOUND_OFFSET;
+  const seLng = longitude + BOUND_OFFSET;
+
+  const req = new Request(
+    `${API_URL}/data.json?opt=1/mAQI/a10/cC5&fetch=true&nwlat=${nwLat}&selat=${seLat}&nwlng=${nwLng}&selng=${seLng}&fields=ID`
+  );
+
+  /** @type {{ code?: number; data?: Array<Array<number>>; fields?: Array<string>;  }} */
+  const res = await req.loadJSON();
+
+  const RATE_LIMIT = 429;
+  if (res.code === RATE_LIMIT) return DEFAULT_SENSOR_ID;
+
+  const { fields, data } = res;
+
+  const sensorIdIndex = fields.indexOf("ID");
+  const latIndex = fields.indexOf("Lat");
+  const lonIndex = fields.indexOf("Lon");
+  const typeIndex = fields.indexOf("Type");
+  const OUTDOOR = 0;
+
+  const [closestSensor] = data
+    .filter((datum) => datum[typeIndex] === OUTDOOR)
+    .sort(
+      (a, b) =>
+        haversine(
+          { latitude, longitude },
+          { latitude: a[latIndex], longitude: a[lonIndex] }
+        ) -
+        haversine(
+          { latitude, longitude },
+          { latitude: b[latIndex], longitude: b[lonIndex] }
+        )
+    );
+
+  return closestSensor ? closestSensor[sensorIdIndex] : DEFAULT_SENSOR_ID;
+}
+
+/**
+ * Returns the haversine distance between start and end.
+ *
+ * @param {LatLon} start
+ * @param {LatLon} end
+ * @returns {number}
+ */
+function haversine(start, end) {
```

Comment:
This function determines the distance between two points using the Haversine formula.

Reply:
Since this is an input into the sort function, should you not precompute the distances once?

SF is *covered* in sensors, so this could be mildly expensive.

Reply:
yeah I wish this API literally would return the nearest sensor but I can't figure that out! There's only so much we can do.